### PR TITLE
DOC: ensure that docstrings for np.ndarray.copy, np.copy and np.array all mention (or link to explanation) that object arrays are shallow-copied

### LINKED ISSUE #25562: BUG: np.array() only shallow-copies arrays with dtype=object
### Describe the issue:

Arrays with `dtype=object` seem to be shallow-copied by `np.array`.
This is surprising for 2 reasons:
- other dtypes are deep-copied
- raw `objects` are not singletons and nothing prevents them to be copied in general, as illustrated hereafter
```python
import copy
a = object()
assert copy.copy(a) is not a # pass
```

### Reproduce the code example:

```python
import numpy as np
a = np.array(object())

# copy=True is also the default behaviour, I'm just making it extra explicit
b = np.array(a, copy=True) 
assert b.item() is not a.item()
```

### Error message:

```shell
AssertionError(...)
```

### Python and NumPy Versions:

1.26.3
3.12.0 (main, Oct 27 2023, 11:50:57) [Clang 15.0.0 (clang-1500.0.40.1)]

### Runtime Environment:

[{'numpy_version': '1.26.3',
  'python': '3.12.0 (main, Oct 27 2023, 11:50:57) [Clang 15.0.0 '
            '(clang-1500.0.40.1)]',
  'uname': uname_result(system='Darwin', node='kwanzaabot.home', release='23.2.0', version='Darwin Kernel Version 23.2.0: Wed Nov 15 21:55:06 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6020', machine='arm64')},
 {'simd_extensions': {'baseline': ['NEON', 'NEON_FP16', 'NEON_VFPV4', 'ASIMD'],
                      'found': ['ASIMDHP'],
                      'not_found': ['ASIMDFHM']}},
 {'architecture': 'armv8',
  'filepath': '/Users/clm/.pyenv/versions/3.12.0/envs/astropy.dev/lib/python3.12/site-packages/numpy/.dylibs/libopenblas64_.0.dylib',
  'internal_api': 'openblas',
  'num_threads': 10,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.23.dev'}]

### Context for the issue:

This came out as the root cause of a bug in astropy (https://github.com/astropy/astropy/issues/13435).

Perhaps a more striking illustration is
```python
a = np.array(1028)
b = np.array(a, copy=True) 
assert b.item() is not a.item() # pass

a = np.array(1028, dtype=object)
b = np.array(a, copy=True) 
assert b.item() is not a.item() # fail
```

### PR DESCRIPTION
**Pull request by neutrinoceros**
This is meant to close #25562
Despite my best effort, I couldn't manage to get a successful build of documentation locally, which I suspect has to do with a circular dependency (`numpy docs`-> `matplotlib` -> `contourpy` -> `numpy < 2.0`), but I'm unsure, so I'm opening as a draft first so I can hopefully inspect CI artifacts before opening this for review.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
Failures seem unrelated and it looks to me that docs are not built through CI on this PR, so I guess I'll just open this for review now.

File: numpy/_core/_add_newdocs.py [hunk 1]
```diff
@@ -810,7 +810,9 @@
         If true (default), then the object is copied.  Otherwise, a copy will
```
Comment (mhvk) [inline numpy/_core/_add_newdocs.py:810]: How about also changing here `then the array data will be copied.`

File: numpy/_core/_add_newdocs.py [hunk 2]
```diff
@@ -810,7 +810,9 @@
...
         only be made if ``__array__`` returns a copy, if obj is a nested
         sequence, or if a copy is needed to satisfy any of the other
-        requirements (``dtype``, ``order``, etc.).
+        requirements (``dtype``, ``order``, etc.). Note that any copy made is
```
Comment (mhvk) [inline numpy/_core/_add_newdocs.py:813]: And then here. 
```
... Note that any copy of
the data is shallow, i.e., for arrays with object dtype, the new
array will point to the same objects. See notes for `copy`.
```

File: numpy/_core/_add_newdocs.py [hunk 3]
```diff
@@ -3344,6 +3347,10 @@
     function :func:`numpy.copy` is similar, but it defaults to using order 'K',
     and will not pass sub-classes through by default.
 
+    Note that np.ndarray.copy is a shallow copy and will not copy object
```
Comment (mhvk) [inline numpy/_core/_add_newdocs.py:3350]: Maybe here
```
The copy made is of the data is shallow, i.e., for arrays with object dtype,
the new array will point to the same objects. See Notes from `copy`.
```

**Review by mhvk — COMMENTED** [review]
Thanks! I have a suggestion inline for making clear that it is the copy of the data that is shallow, not of the array (that would be a `view`, I guess). What do you think?

Another thing I wondered about is, if `ndarray.copy()` is meant to be the preferred method, the actual example/warning should be there rather than in the docstring for `np.copy`. But perhaps best to just keep the changes minimal.

File: numpy/_core/_add_newdocs.py [hunk 4]
```diff
@@ -810,7 +810,9 @@
         If true (default), then the object is copied.  Otherwise, a copy will
         only be made if ``__array__`` returns a copy, if obj is a nested
         sequence, or if a copy is needed to satisfy any of the other
-        requirements (``dtype``, ``order``, etc.).
+        requirements (``dtype``, ``order``, etc.). Note that any copy made is
+        shallow, which is mainly important for arrays containing Python
+        objects. See Notes from `copy`.
```
Comment (rkern) [inline numpy/_core/_add_newdocs.py:815]: The `Notes` of `np.copy` don't mention anything about this. It's only in the `Examples`. I'd probably move or copy those examples over to `ndarray.copy` as the canonical API for copying (`np.copy` refers to `ndarray.copy` as the preferred method).

I see that you copied the verbiage from `np.copy`, but I think that wording is misleading. I would dispute the idea that it's always a "shallow copy" that is "mainly important for arrays containing Python objects". If it's not an array that contains Python objects, the distinction between deep and shallow doesn't even exist. That wording raises a question what a "shallow copy" even means for the other dtypes. I'd prefer to flip it around: "For arrays containing Python objects (e.g. ``dtype=object``), the copy is a shallow one."

**Comment by neutrinoceros** [comment]
Thank you both for your feedback. I moved the bulk of the explanation (narrative examples) from `np.copy` to the preferred implementation `np.ndarray.copy` and rephrased as requested.

**Review by mhvk — APPROVED** [review]
Nice! Approving, though with things like documentation, two pairs of eyes are generally better, so maybe @rkern can have another look too.

**Review by rkern — APPROVED** [review]

**Comment by mhvk** [comment]
Two approvals seems good enough - let's get this in!

**Pull request merged**